csrc/id_model/indexing.cpp

```diff
@@ -210,8 +210,16 @@ class AllocationDomainSetup : private kir::IrVisitor {
     }
 
     if (use_set_allocation_domain) {
-      allocation_domains = tv->getAllocationDomain();
-      contiguity = tv->domain()->contiguity();
+      std::unordered_set<IterDomain*> ca_ids(
+          tv->getLoopDomain().begin(),
+          tv->getLoopDomain().begin() + tv->getComputeAtPosition());
```

Comment:
Isn't this necessarily correct with gmem tensors? Not common to have non-input gmem tensors, though.

Reply:
Interesting question. Let's say we have a fusion:
```
T1g[3, 5] ca_pos(1) = set(T0g[3, 5]);
T2g[3, 5] = set(T1g[3, 5]);
```
what kernel do we want to have? Two options are:

Option 1:

```C++
// T1g allocated as [5]
for (auto i : irange(3)) {
  for (auto j : irange(5)) {
    T1g[j] = T0g[i, j];
  }
  for (auto j : irange(5)) {
    T2g[i, j] = T1g[j];
  }
  grid_sync();
}
```

Option 2:

```C++
// T1g allocated as [3, 5]
for (auto i : irange(3)) {
  for (auto j : irange(5)) {
    T1g[i, j] = T0g[i, j];
  }
  for (auto j : irange(5)) {
    T2g[i, j] = T1g[i, j];
  }
  // no grid sync
}
```

I personally prefers option 1, because if we want to allocate `T1g` as `[3, 5]`, why not just don't inline `T1g` at all?

Reply:
Oh, are you saying that I should add a `!ir_utils::isMemorySharedAcross(tv->getMemoryType(), pt)` check?

Reply:
I'm not saying which option is preferred, but since option 2 is what we do right now, the allocation domain should be `[3, 5]`. Otherwise, we would need to update other things too.

> Oh, are you saying that I should add a !ir_utils::isMemorySharedAcross(tv->getMemoryType(), pt) check?

I'm not sure.  Allocation domain inference is non trivial. 

What is the right allocation domain and what is the current inferred allocation domain?

Reply:
> I'm not saying which option is preferred, but since option 2 is what we do right now, the allocation domain should be `[3, 5]`. 

Then let me use the whole allocation domain for gmem tensor.

> What is the right allocation domain and what is the current inferred allocation domain?

I don't think `isMemorySharedAcross` is related to this PR.

For this PR, the problematic allocation domain is the one of this tensor:

```C++
T11_s[ iS61{i0}, iblockIdx.x65{( ceilDiv(i3, 128) )}, iblockIdx.y67{( ceilDiv(i4, 128) )}, iblockIdx.z64{2}, ithreadIdx.z311{( ceilDiv(( ( ceilDiv(128, 4) ) * 4 ), 64) )}, ithreadIdx.y319{( ceilDiv(( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 ), 64) )}, iS321{( ceilDiv(64, 16) )}, iS323{( ceilDiv(64, 8) )}, ithreadIdx.x334{( ( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 ) * ( ceilDiv(8, 2) ) )}, iS330{( ceilDiv(8, 8) )}, iS328{2}, iV333{2} ] ca_pos( 3 ) produce_pos( 4 )
 logical domain : (iS61{i0}, iS62{i3}, iS63{i4}, iblockIdx.z64{2})
 allocation domain : (iS61{i0}, iblockIdx.x65{( ceilDiv(i3, 128) )}, iblockIdx.y67{( ceilDiv(i4, 128) )}, iblockIdx.z64{2}, iS304{( ceilDiv(128, 4) )}, iS306{4}, iS302{( ceilDiv(( ceilDiv(128, 8) ), 4) )}, iS307{4}, iS301{8})
 contiguity: t t t t t t t t t
  Split: iS62{i3} by factor 128 -> iblockIdx.x65{( ceilDiv(i3, 128) )}, iS66{128}
  Split: iS63{i4} by factor 128 -> iblockIdx.y67{( ceilDiv(i4, 128) )}, iS68{128}
  Split: iS66{128} by factor 4 -> iS308{( ceilDiv(128, 4) )}, iS309{4}
  Merge: iS308{( ceilDiv(128, 4) )} and iS309{4} -> iS310{( ( ceilDiv(128, 4) ) * 4 )}
  Split: iS310{( ( ceilDiv(128, 4) ) * 4 )} by factor 64 -> ithreadIdx.z311{( ceilDiv(( ( ceilDiv(128, 4) ) * 4 ), 64) )}, iS312{64}
  Split: iS68{128} by factor 8 -> iS313{( ceilDiv(128, 8) )}, iS314{8}
  Split: iS313{( ceilDiv(128, 8) )} by factor 4 -> iS315{( ceilDiv(( ceilDiv(128, 8) ), 4) )}, iS316{4}
  Merge: iS315{( ceilDiv(( ceilDiv(128, 8) ), 4) )} and iS316{4} -> iS317{( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 )}
  Merge: iS317{( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 )} and iS314{8} -> iS318{( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 )}
  Split: iS318{( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 )} by factor 64 -> ithreadIdx.y319{( ceilDiv(( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 ), 64) )}, iS320{64}
  Split: iS312{64} by factor 16 -> iS321{( ceilDiv(64, 16) )}, iS322{16}
  Split: iS320{64} by factor 8 -> iS323{( ceilDiv(64, 8) )}, iS324{8}
  Split: iS322{16} by factor 8 -> iS325{( ceilDiv(16, 8) )}, iS326{8}
  Split: iS325{( ceilDiv(16, 8) )} by factor 2 -> iS327{( ceilDiv(( ceilDiv(16, 8) ), 2) )}, iS328{2}
  Merge: iS327{( ceilDiv(( ceilDiv(16, 8) ), 2) )} and iS326{8} -> iS329{( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 )}
  Split: iS324{8} by factor 8 -> iS330{( ceilDiv(8, 8) )}, iS331{8}
  Split: iS331{8} by factor 2 -> iS332{( ceilDiv(8, 2) )}, iV333{2}
  Merge: iS329{( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 )} and iS332{( ceilDiv(8, 2) )} -> ithreadIdx.x334{( ( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 ) * ( ceilDiv(8, 2) ) )}
 loop domain : (iS61{i0}, iblockIdx.x65{( ceilDiv(i3, 128) )}, iblockIdx.y67{( ceilDiv(i4, 128) )}, iblockIdx.z64{2}, ithreadIdx.z311{( ceilDiv(( ( ceilDiv(128, 4) ) * 4 ), 64) )}, ithreadIdx.y319{( ceilDiv(( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 ), 64) )}, iS321{( ceilDiv(64, 16) )}, iS323{( ceilDiv(64, 8) )}, ithreadIdx.x334{( ( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 ) * ( ceilDiv(8, 2) ) )}, iS330{( ceilDiv(8, 8) )}, iS328{2}, iV333{2})
```

and the change I am doing here is to remove this `iS61{i0}` from allocation domain. This IterDomain is serial so it has nothing to do with `isMemorySharedAcross`.

But I believe in general, we still need to do `isMemorySharedAcross`, for example, if we have allocation domain

```C++
TIDx, 5
```

for a smem tensor, and `TIDx` is inlined in the loop domain, for this case, we still should keep this `TIDx`.

Reply:
> Then let me use the whole allocation domain for gmem tensor.

But in the future, I believe we need to remove this special case, because in the world of multi GPU, global memory is no different than shared memory. They are both a "shared memory" on a certain level.

Reply:
> > Then let me use the whole allocation domain for gmem tensor.
> 
> But in the future, I believe we need to remove this special case, because in the world of multi GPU, global memory is no different than shared memory. They are both a "shared memory" on a certain level.

Of course, but then at that point, inferring allocation domains would be even more fragile. Something would need to be done by the schedulers.

Reply:
> > I'm not saying which option is preferred, but since option 2 is what we do right now, the allocation domain should be `[3, 5]`.
> 
> Then let me use the whole allocation domain for gmem tensor.
> 
> > What is the right allocation domain and what is the current inferred allocation domain?
> 
> I don't think `isMemorySharedAcross` is related to this PR.
> 
> For this PR, the problematic allocation domain is the one of this tensor:
> 
> ```c++
> T11_s[ iS61{i0}, iblockIdx.x65{( ceilDiv(i3, 128) )}, iblockIdx.y67{( ceilDiv(i4, 128) )}, iblockIdx.z64{2}, ithreadIdx.z311{( ceilDiv(( ( ceilDiv(128, 4) ) * 4 ), 64) )}, ithreadIdx.y319{( ceilDiv(( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 ), 64) )}, iS321{( ceilDiv(64, 16) )}, iS323{( ceilDiv(64, 8) )}, ithreadIdx.x334{( ( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 ) * ( ceilDiv(8, 2) ) )}, iS330{( ceilDiv(8, 8) )}, iS328{2}, iV333{2} ] ca_pos( 3 ) produce_pos( 4 )
>  logical domain : (iS61{i0}, iS62{i3}, iS63{i4}, iblockIdx.z64{2})
>  allocation domain : (iS61{i0}, iblockIdx.x65{( ceilDiv(i3, 128) )}, iblockIdx.y67{( ceilDiv(i4, 128) )}, iblockIdx.z64{2}, iS304{( ceilDiv(128, 4) )}, iS306{4}, iS302{( ceilDiv(( ceilDiv(128, 8) ), 4) )}, iS307{4}, iS301{8})
>  contiguity: t t t t t t t t t
>   Split: iS62{i3} by factor 128 -> iblockIdx.x65{( ceilDiv(i3, 128) )}, iS66{128}
>   Split: iS63{i4} by factor 128 -> iblockIdx.y67{( ceilDiv(i4, 128) )}, iS68{128}
>   Split: iS66{128} by factor 4 -> iS308{( ceilDiv(128, 4) )}, iS309{4}
>   Merge: iS308{( ceilDiv(128, 4) )} and iS309{4} -> iS310{( ( ceilDiv(128, 4) ) * 4 )}
>   Split: iS310{( ( ceilDiv(128, 4) ) * 4 )} by factor 64 -> ithreadIdx.z311{( ceilDiv(( ( ceilDiv(128, 4) ) * 4 ), 64) )}, iS312{64}
>   Split: iS68{128} by factor 8 -> iS313{( ceilDiv(128, 8) )}, iS314{8}
>   Split: iS313{( ceilDiv(128, 8) )} by factor 4 -> iS315{( ceilDiv(( ceilDiv(128, 8) ), 4) )}, iS316{4}
>   Merge: iS315{( ceilDiv(( ceilDiv(128, 8) ), 4) )} and iS316{4} -> iS317{( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 )}
>   Merge: iS317{( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 )} and iS314{8} -> iS318{( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 )}
>   Split: iS318{( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 )} by factor 64 -> ithreadIdx.y319{( ceilDiv(( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 ), 64) )}, iS320{64}
>   Split: iS312{64} by factor 16 -> iS321{( ceilDiv(64, 16) )}, iS322{16}
>   Split: iS320{64} by factor 8 -> iS323{( ceilDiv(64, 8) )}, iS324{8}
>   Split: iS322{16} by factor 8 -> iS325{( ceilDiv(16, 8) )}, iS326{8}
>   Split: iS325{( ceilDiv(16, 8) )} by factor 2 -> iS327{( ceilDiv(( ceilDiv(16, 8) ), 2) )}, iS328{2}
>   Merge: iS327{( ceilDiv(( ceilDiv(16, 8) ), 2) )} and iS326{8} -> iS329{( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 )}
>   Split: iS324{8} by factor 8 -> iS330{( ceilDiv(8, 8) )}, iS331{8}
>   Split: iS331{8} by factor 2 -> iS332{( ceilDiv(8, 2) )}, iV333{2}
>   Merge: iS329{( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 )} and iS332{( ceilDiv(8, 2) )} -> ithreadIdx.x334{( ( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 ) * ( ceilDiv(8, 2) ) )}
>  loop domain : (iS61{i0}, iblockIdx.x65{( ceilDiv(i3, 128) )}, iblockIdx.y67{( ceilDiv(i4, 128) )}, iblockIdx.z64{2}, ithreadIdx.z311{( ceilDiv(( ( ceilDiv(128, 4) ) * 4 ), 64) )}, ithreadIdx.y319{( ceilDiv(( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 ), 64) )}, iS321{( ceilDiv(64, 16) )}, iS323{( ceilDiv(64, 8) )}, ithreadIdx.x334{( ( ( ceilDiv(( ceilDiv(16, 8) ), 2) ) * 8 ) * ( ceilDiv(8, 2) ) )}, iS330{( ceilDiv(8, 8) )}, iS328{2}, iV333{2})
> ```
> 
> and the change I am doing here is to remove this `iS61{i0}` from allocation domain. This IterDomain is serial so it has nothing to do with `isMemorySharedAcross`.
> 
> But I believe in general, we still need to do `isMemorySharedAcross`, for example, if we have allocation domain
> 
> ```c++
> TIDx, 5
> ```
> 
> for a smem tensor, and `TIDx` is inlined in the loop domain, for this case, we still should keep this `TIDx`.

I see. It seems that we would need to know which iter domains of the allocation domain are inlined or not. That information is readily available for the loop domain but not for the allocation domain, but without that information, it doesn't seem to find which domains can be dropped or not.

Can we introduce some quick WAR for this pattern?

Reply:
> Can we introduce some quick WAR for this pattern?

What kind of WAR are you thinking about? Doesn't the approach in this PR makes sense? My feeling for what we should do is, for inlined tensor, the allocation domain and the loop domain should both contain the inlined part. The allocation domain and loop domain can only be different for the non-inlined part.

Reply:
Oh, I didn't notice the code was updated. Let me take a look again.

Reply:
The updated code makes sense to me. I just left a suggestion.